File: app/src/main/java/net/ducksmanager/whattheduck/IssueEdgeAdapter.java
package net.ducksmanager.whattheduck;

import android.app.Activity;
import android.content.res.Configuration;
import android.view.View;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueDetails;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import static net.ducksmanager.whattheduck.WhatTheDuck.*;

public class IssueEdgeAdapter extends ItemAdapter<InducksIssueWithUserIssueDetails> {
    private final int orientation;
    private final RecyclerView recyclerView;
    private Integer expectedEdgeHeight;

    IssueEdgeAdapter(ItemList itemList, List<InducksIssueWithUserIssueDetails> items, RecyclerView recyclerView, int orientation) {
        super(itemList, R.layout.row_edge, items);
        this.orientation = orientation;
        this.recyclerView = recyclerView;
    }

    @Override
    protected ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    @Override
    protected View.OnClickListener getOnClickListener() {
        return null;
    }

    class ViewHolder extends ItemAdapter.ViewHolder {
        final ImageView edgeImage;

        ViewHolder(View v) {
            super(v);
            edgeImage = v.findViewById(R.id.edgeimage);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull ItemAdapter.ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);

        ViewHolder itemHolder = (ViewHolder) holder;

        InducksIssueWithUserIssueDetails i = getItem(position);
        if (i != null) {
            if (expectedEdgeHeight == null) {
                expectedEdgeHeight = orientation == Configuration.ORIENTATION_LANDSCAPE ? recyclerView.getHeight() : recyclerView.getWidth();
            }

            Picasso
                .with(((ViewHolder) holder).itemView.getContext())
                .load(getEdgeUrl(i))
                .resize(0, expectedEdgeHeight)
                .rotate(orientation == Configuration.ORIENTATION_LANDSCAPE ? 0 : 90f)
                .into(itemHolder.edgeImage);
        }
    }

    @NonNull
    private String getEdgeUrl(InducksIssueWithUserIssueDetails i) {
        return String.format(
            "%s/edges/%s/gen/%s.%s.png",
            config.getProperty(CONFIG_KEY_EDGES_URL),
            selectedCountry,
            selectedPublication
                .replaceFirst("[^/]+/", "")
                .replaceAll(" ", ""),
            i.getIssue().getInducksIssueNumber().replaceAll(" ", ""));
    }

    @Override
    protected boolean isPossessed(InducksIssueWithUserIssueDetails i) {
        return i.getUserIssue() != null;
    }

    @Override
    protected Integer getPrefixImageResource(InducksIssueWithUserIssueDetails i, Activity activity) {
        return null;
    }

    @Override
    protected Integer getSuffixImageResource(InducksIssueWithUserIssueDetails i) {
        return null;
    }

    @Override
    protected String getSuffixText(InducksIssueWithUserIssueDetails i) {
        return null;
    }

    @Override
    protected String getIdentifier(InducksIssueWithUserIssueDetails i) {
        return null;
    }

    @Override
    protected String getText(InducksIssueWithUserIssueDetails i) {
        return null;
    }

    @Override
    protected String getComparatorText(InducksIssueWithUserIssueDetails i) {
        return i.getIssue().getInducksIssueNumber();
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/IssueEdgeAdapter.kt
package net.ducksmanager.whattheduck

import android.app.Activity
import android.content.res.Configuration
import android.view.View
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import net.ducksmanager.persistence.models.composite.InducksIssueWithUserIssueDetails

class IssueEdgeAdapter internal constructor(
    itemList: ItemList<InducksIssueWithUserIssueDetails>,
    items: List<InducksIssueWithUserIssueDetails>,
    private val recyclerView: RecyclerView,
    private val orientation: Int
) : ItemAdapter<InducksIssueWithUserIssueDetails>(itemList, R.layout.row_edge, items) {
    private var expectedEdgeHeight: Int? = null

    override fun getViewHolder(v: View?) = ViewHolder(v)


    override val onClickListener: View.OnClickListener?
        get() = null

    inner class ViewHolder(v: View?) : ItemAdapter<InducksIssueWithUserIssueDetails>.ViewHolder(v!!) {
        val edgeImage: ImageView = v!!.findViewById(R.id.edgeimage)

    }

    override fun onBindViewHolder(holder: ItemAdapter<InducksIssueWithUserIssueDetails>.ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val itemHolder = holder as ViewHolder
        val item = getItem(position)

        if (expectedEdgeHeight == null) {
            expectedEdgeHeight = if (orientation == Configuration.ORIENTATION_LANDSCAPE) recyclerView.height else recyclerView.width
        }
        Picasso
            .with(holder.itemView.getContext())
            .load(getEdgeUrl(item))
            .resize(0, expectedEdgeHeight!!)
            .rotate(if (orientation == Configuration.ORIENTATION_LANDSCAPE) 0f else 90f)
            .into(itemHolder.edgeImage)
    }

    private fun getEdgeUrl(i: InducksIssueWithUserIssueDetails): String {
        return String.format(
            "%s/edges/%s/gen/%s.%s.png",
            WhatTheDuck.config.getProperty(WhatTheDuck.CONFIG_KEY_EDGES_URL),
            WhatTheDuck.selectedCountry,
            WhatTheDuck.selectedPublication!!
                .replaceFirst("[^/]+/".toRegex(), "")
                .replace(" ".toRegex(), ""),
            i.issue.inducksIssueNumber.replace(" ".toRegex(), ""))
    }

    override fun isPossessed(item: InducksIssueWithUserIssueDetails): Boolean {
        return item.userIssue != null
    }

    override fun getPrefixImageResource(i: InducksIssueWithUserIssueDetails, activity: Activity): Int? {
        return null
    }

    override fun getSuffixImageResource(i: InducksIssueWithUserIssueDetails): Int? {
        return null
    }

    override fun getSuffixText(i: InducksIssueWithUserIssueDetails): String? {
        return null
    }

    override fun getIdentifier(i: InducksIssueWithUserIssueDetails): String? {
        return null
    }

    override fun getText(i: InducksIssueWithUserIssueDetails): String? {
        return null
    }

    override fun getComparatorText(i: InducksIssueWithUserIssueDetails): String? {
        return i.issue.inducksIssueNumber
    }

}